File: plugins/ios/chat-module/Sources/StarterChat/Models/ChatMessage.swift
import Foundation
import StarterAI
import SwiftData

@Model
public final class ChatMessage {
    public var id: UUID
    public var sessionId: UUID
    public var role: String
    public var content: String
    public var createdAt: Date
    public var syncedAt: Date?

    public init(id: UUID = UUID(), sessionId: UUID, role: String, content: String, createdAt: Date = Date(), syncedAt: Date? = nil) {
        self.id = id
        self.sessionId = sessionId
        self.role = role
        self.content = content
        self.createdAt = createdAt
        self.syncedAt = syncedAt
    }

    public convenience init(from message: AIMessage, sessionId: UUID) {
        self.init(
            id: message.id,
            sessionId: sessionId,
            role: message.role.rawValue,
            content: message.content,
            createdAt: message.createdAt
        )
    }
}

File: plugins/flutter/chat/lib/src/models/chat_message.dart
import 'package:starter_ai/starter_ai.dart';

/// Mirrors `ChatMessage.swift` (the SwiftData `@Model`). Here it is a plain
/// immutable value object; persistence is handled by `ChatStore`'s drift
/// table, which stores the same fields.
class ChatMessage {
  const ChatMessage({
    required this.id,
    required this.sessionId,
    required this.role,
    required this.content,
    required this.createdAt,
    this.syncedAt,
  });

  final String id;
  final String sessionId;
  final String role;
  final String content;
  final DateTime createdAt;
  final DateTime? syncedAt;

  /// Mirrors `ChatMessage.init(from: AIMessage, sessionId:)`.
  factory ChatMessage.fromAIMessage(AIMessage message, {required String sessionId}) {
    return ChatMessage(
      id: message.id,
      sessionId: sessionId,
      role: message.role.wireValue,
      content: message.content,
      createdAt: message.createdAt,
    );
  }

  ChatMessage copyWith({DateTime? syncedAt}) => ChatMessage(
    id: id,
    sessionId: sessionId,
    role: role,
    content: content,
    createdAt: createdAt,
    syncedAt: syncedAt ?? this.syncedAt,
  );
}
